### ISSUE TITLE
nv_gpu_memory_used_bytes metric does not decrease on model unload

### ISSUE THREAD

**Issue by cliveseldon** [NVIDIA/tensorrt-inference-server#1082]
From local testing with 20.01 the metric nv_gpu_memory_used_bytes exposed on `/metrics` does not  decrease on model unload. Assuming this is expected would there be some way to expose the actual memory used by the loaded models?

I ask as orchestrators (like when running inside  kubernetes) might wish to determine the available memory for new models on the server.

**Comment by nieksand** [comment]
Which framework does your model use?  Tensorflow?

**Comment by cliveseldon** [comment]
I was just running the test "simple" and "resnet" models from the examples.

**Comment by GuanLuo** [comment]
"simple" is a TF model, so that may explain. TensorFlow will allocate GPU memory as it needs, but it never releases them until the process exit, so it is not a good target for testing this metric. Empirically, TensorRT and ONNX frameworks are good at releasing GPU resources on model unload. PyTorch and Caffe2 are good too, but as far as I know, they have some issues (#992) in multiple instance case.

**Comment by cliveseldon** [comment]
Thanks. Good to know there is a difference. So this would pose an issue for using TRTIS more as a service to deploy arbitrary models in kubernetes.

**Comment by deadeyegoodwin** [comment]
Any deployment plan should be aware of the limitations of the different backends. TF is not a particularly good framework for inference for a couple of reasons (this memory usage policy being one of them). If you want to have a long-running "dynamic model repository" TRTIS instance where you use the model-control APIs to load and unload models, then you need to account for TF. TRTIS provides a command-line option to limit TF to a fraction of the GPU memory... but this doesn't cause it to release its memory. 

For a system like kubernetes an alternative is to use a "static model repository" TRTIS instance where TRTIS is started/initialized with some set of models and that set of models never changes. If a change is needed then use k8s to change configuration via a rolling update or similar.

Each approach has advantages and disadvantages.

**Comment by cliveseldon** [comment]
Thanks for your feedback. Very useful.
